File: units/Units.go
package units

import (
	"fmt"
)

type Price struct {
	Metal     int
	Crystal   int
	Deuterium int
}

func (p *Price) Add(price Price) {
	p.Metal += price.Metal
	p.Crystal += price.Crystal
	p.Deuterium += price.Deuterium
}

type CombatUnit struct {
	Price            Price
	OgameID          int
	Name             string
	HullPlating      float64
	Shield           float64
	ShieldPower      float64
	WeaponPower      float64
	ArmourTechnology int
	WeaponTechnology int
	ShieldTechnology int
	RapidfireFrom    map[string]int
	RapidfireAgainst map[string]int
}

func (ship *CombatUnit) IsDead() bool {
	return ship.HullPlating <= 0
}

func (ship *CombatUnit) IsAlive() bool {
	return ship.HullPlating > 0
}

func (ship *CombatUnit) GetPrice() Price {
	return ship.Price
}

func (ship *CombatUnit) GetRapidfireAgainst(name string) int {
	return ship.RapidfireAgainst[name]
}

func (ship *CombatUnit) InitHullPlating() {
	ship.HullPlating = (1 + (float64(ship.ArmourTechnology) / 10)) * ((float64(ship.Price.Metal) + float64(ship.Price.Crystal)) / 10)
}

func (ship *CombatUnit) GetInitialHullPlating() int {
	return (1 + (ship.ArmourTechnology / 10)) * ((ship.Price.Metal + ship.Price.Crystal) / 10)
}

func (ship *CombatUnit) GetHullPlating() float64 {
	return ship.HullPlating
}

func (ship *CombatUnit) SetHullPlating(hullPlating float64) {
	ship.HullPlating = hullPlating
}

func (ship *CombatUnit) GetName() string {
	return ship.Name
}

func (ship *CombatUnit) GetWeapon() float64 {
	return ship.WeaponPower
}

func (ship *CombatUnit) GetShield() float64 {
	return ship.Shield
}

func (ship *CombatUnit) SetShield(shield float64) {
	ship.Shield = shield
}

func (ship *CombatUnit) RestoreShield() {
	ship.Shield = ship.ShieldPower * (1 + 0.1*0)
}

func (ship *CombatUnit) String() string {
	return fmt.Sprintf("%s with %d:%d:%d", ship.GetName(), ship.GetHullPlating(), ship.GetShield(), ship.GetWeapon())
}

File: units/Units.py
class Price(object):
    Metal = 0
    Crystal = 0
    Deuterium = 0

    def __init__(self, metal=0, crystal=0, deut=0):
        self.Metal = metal
        self.Crystal = crystal
        self.Deuterium = deut

    def Add(self, price):
        self.Metal += price.Metal
        self.Crystal += price.Crystal
        self.Deuterium += price.Deuterium


class CombatUnit(object):
    Price = Price()
    OgameID = 0
    Name = ""
    HullPlating = 0
    Shield = 0
    ShieldPower = 0
    WeaponPower = 0
    ArmourTechnology = 0
    WeaponTechnology = 0
    ShieldTechnology = 0
    RapidfireFrom = {}
    RapidfireAgainst = {}

    def IsDead(self):
        return self.HullPlating <= 0

    def IsAlive(self):
        return self.HullPlating > 0

    def GetPrice(self):
        return self.Price

    def GetRapidfireAgainst(self, name):
        return self.RapidfireAgainst.get(name, 0)

    def InitHullPlating(self):
        self.HullPlating = (1 + (float(self.ArmourTechnology) / 10)) * ((float(self.Price.Metal) + float(self.Price.Crystal)) / 10)

    def GetInitialHullPlating(self):
        return (1 + (self.ArmourTechnology / 10)) * ((self.Price.Metal + self.Price.Crystal) / 10)

    def GetHullPlating(self):
        return self.HullPlating

    def SetHullPlating(self, hullPlating):
        self.HullPlating = hullPlating

    def GetName(self):
        return self.Name

    def GetWeapon(self):
        return self.WeaponPower

    def GetShield(self):
        return self.Shield

    def SetShield(self, shield):
        self.Shield = shield

    def RestoreShield(self):
        self.Shield = self.ShieldPower * (1 + 0.1*0)

    def String(self):
        return "%s with %d:%d:%d" % (self.GetName(), self.GetHullPlating(), self.GetShield(), self.GetWeapon())
